Safer erc20 (#4524)
ethereum/EIPs: EIPS/eip-4524.md

---
eip: 4524
title: Safer ERC20

> description: Extending ERC20 with ERC165 and adding safeTransfer (like ERC721 and ERC1155)

Reviewer: ```suggestion
description: Extending ERC-20 with ERC-165 and adding safeTransfer (like ERC-721 and ERC-1155)
```

author: William Schwab (@wschwab)
discussions-to: https://ethereum-magicians.org/t/why-isnt-there-an-erc-for-safetransfer-for-erc20/7604
status: Draft
type: Standards Track
category: ERC
created: 2021-12-05
requires: 20, 165
---

## Abstract

> This standard extends [ERC-20](./erc-20.md) tokens with [EIP-165](./eip-165.md), and adds familiar functions from [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md) ensuring receiving contracts have implemented proper functionality.

Reviewer: ```suggestion
This standard extends [ERC-20](./eip-20.md) tokens with [EIP-165](./eip-165.md), and adds familiar functions from [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md) ensuring receiving contracts have implemented proper functionality.
```

## Specification

> The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

Reviewer 2: ```suggestion
The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
```

Reviewer: Per RFC 2119, it is "key words" https://www.ietf.org/rfc/rfc2119.txt

Author: there's a history here of splitting it into two words, though I see your point

  function safeTransfer(address to, uint256 amount) external returns(bool);
  function safeTransfer(address to, uint256 amount, bytes memory data) external returns(bool);
  function safeTransferFrom(address from, address to, uint256 amount) external returns(bool);
  function safeTransferFrom(address from, address to, uint256 amount, bytes memory data) external returns(bool);
}
```
`safeTransfer` and `safeTransferFrom` MUST transfer as expected to EOA addresses, and to contracts implementing `ERC20Receiver` and returning the function selector (`0x4fc35859`) when called, and MUST revert when transferring to a contract which either does not have `ERC20Receiver` implemented, or does not return the function selector when called.

In addition, a contract accepting safe transfers MUST implement the following if it wishes to accept safe transfers, and MUST return the function selector (`0x4fc35859`):
```solidity
pragma solidity 0.8.10;

import './IERC165.sol';

interface ERC20Receiver is IERC165 {
  function onERC20Received(
    address _operator,
    address _from,
    uint256 _amount,
    bytes _data
  ) external returns(bytes4);
}
```

## Rationale

> This EIP is meant to be minimal and straightforward. Adding EIP-165 to ERC20 is useful for a number of applications, and outside of a minimal amount of code increasing contract size, carries no downside. The `safeTransfer` and `safeTransferFrom` functions are well recognized from ERC721 and ERC1155, and therefore keeping identical naming conventions is reasonable, and the benefits of being able to check for implementation before transferring are as useful for ERC20 tokens as they are for ERC721 and ERC1155.

Reviewer 2: ```suggestion
This EIP is meant to be minimal and straightforward. Adding EIP-165 to ERC20 is useful for a number of applications, and outside a minimal amount of code increasing contract size, carries no downside. The `safeTransfer` and `safeTransferFrom` functions are well recognized from ERC721 and ERC1155, and therefore keeping identical naming conventions is reasonable, and the benefits of being able to check for implementation before transferring are as useful for ERC20 tokens as they are for ERC721 and ERC1155.
```

Author: unless there's a strong need, I think this should be left as-is

Reviewer 2: Just regular proofreading.
I will mostly skip it from next time, unless absolutely necessary!

Reviewer 3: @wschwab You think it should be left as is because you believe that "...and outside a minimal amount..." is more correct, or for some other reason?

FWIW, I agree with @mryalamanchi that "...and outside of a minimal amount..." is grammatically correct.

Author: I;m of the opinion that my version was closer to common spoken parlance, though that's subjective, and once I've got multiple opinions asking me to change it, I'll change it in the next push.

Reviewer 3: Perhaps I'm just getting old and not keeping up with language evolution.  😖  We could do a poll of teen aged girls to see which sounds more correct to them, since they are the dominant drivers of language evolution in human society.  😄

Reviewer 2: I was coming at it from the POV of regular proofreading. Even a small preposition can sometimes make a huge diff (remember those tricky word problems to be written in code?).

> Another easy backport from EIP721 and EIP1155 might be the inclusion of a metadata URI for tokens, allowing them to easily reference logo and other details. This has not been included both in order to keep this EIP as minimal as possible, and because it is already sufficiently covered by [EIP-1046](./eip-1046.md).

Reviewer 2: ```suggestion
Another easy backport from EIP721 and EIP1155 might be the inclusion of a metadata URI for tokens, allowing them to easily reference logo and other details. This has not been included, both in order to keep this EIP as minimal as possible, and because it is already sufficiently covered by [EIP-1046](./eip-1046.md).
```
